### PR TITLE
feature(@nestjs/core) display known dependency names in unknown dependency exception

### PR DESCRIPTION
**Pull request by BrunnerLivio**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type

This feature displays the found dependency names when any unknown
dependency is getting injected. Before this feature the error message
gets displayed like (?, +). After this implementation, like this e.g.
(?, CatsService).

New Error Message

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #223

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Looks like a quite useful feature. Thanks @BrunnerLivio, I'll review (and hopefully merge) very soon :)

File: packages/core/errors/messages.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-  length: number,
+  args: any[],
 ) => {
   let message = `Nest can't resolve dependencies of the ${type}`;
   message += ` (`;
 
-  const args = new Array(length).fill('+');
-  args[index] = '?';
-  message += args.join(', ');
+  const dependencies = args.map(arg => arg.name);
```
[the remark this answers is not in the document]
Comment (kamilmysliwiec) [reply packages/core/errors/messages.ts]: Yeah, I'm getting weird messages once the `@Inject()` decorator is used. Besides, when a type is not specified, `Object` is shown instead. We need to test this feature against these scenarios

[the remark this answers is not in the document]
Comment (BrunnerLivio) [reply packages/core/errors/messages.ts]: Yes, I thought things like that could potentially be an issue.
I’ll investigate more onto this and I try to append some more test cases for these cases. So if some edge cases come to your mind like `@Inject()` I would appreciate it if you write them down so I can test them.

I’ll be quite busy this week and maybe also don’t find any time next week. Hope thats not a problem :)

[the remark this answers is not in the document]
Comment (kamilmysliwiec) [reply packages/core/errors/messages.ts]: Not an issue at all :)

**Comment by BrunnerLivio** [comment]
It should work now with `@Inject`.

```TypeScript

constructor (private notKnown: CatService, @Inject('MY_TOKEN') private token: string)

```

throws the error `...(?, MY_TOKEN)...`.

I've also added more test cases, but I can not get them successful. On my local machine the tests are successful, but inside the CI pipeline it seems to fail. The pipeline does not return what the actual issue is as far as I know. Do you know how I can proceed debugging?

**Comment by BrunnerLivio** [comment]
@kamilmysliwiec can you review it again?

File: packages/core/errors/messages.ts [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
-  length: number,
+  args: any[],
 ) => {
   let message = `Nest can't resolve dependencies of the ${type}`;
   message += ` (`;
 
-  const args = new Array(length).fill('+');
-  args[index] = '?';
-  message += args.join(', ');
+  const dependencies = args.map(arg => getDependencyName(arg));
```
Comment (wbhob) [inline packages/core/errors/messages.ts]: this can be shortened to `args.map(getDependencyName)`

**Comment by BrunnerLivio** [comment]
> All over the place, you use the name args. This is not descriptive. Please use dependencies or deps, or something that describes the property.

@wbhob 
Yes, thats because initially in `injector.ts` [line 131](https://github.com/nestjs/nest/pull/814/files#diff-03695d7f859ce8dbde7678750e553570R131) it got called args, so I just went if it. But I can change it to a more descriptive variable name. Also everytime `args` is in type `any` which is not so good... Should I change that too? And if so, in which directory should I place the file of this type, so it is consistent with the general structure oft nest?
m

**Comment by BrunnerLivio** [comment]
@wbhob done

**Comment by wbhob** [comment]
Looks great! Thanks for the contribution.

**Comment by kamilmysliwiec** [comment]
Will review it shortly. Thanks @BrunnerLivio!

**Comment by kamilmysliwiec** [comment]
Thank you for your contribution. Amazing PR! Merging 🔥

**Comment by BrunnerLivio** [comment]
@kamilmysliwiec Thanks, looking forward for more contributions! :)
